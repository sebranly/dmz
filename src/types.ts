export enum Color {
  Green = 'green',
  Yellow = 'yellow',
  Blue = 'blue',
  Pink = 'pink'
}

export enum TimeUnit {
  Hour = 'hour',
  Minute = 'minute',
  Second = 'second'
}

export type TimeValue = {
  [TimeUnit.Hour]: number;
  [TimeUnit.Minute]: number;
  [TimeUnit.Second]: number;
};

export type Timer = {
  timerIndex: number;
  durationSec: number;
  timestampStart: number;
};

export enum Sort {
  oldestToNewest = 'Oldest to Newest',
  newestToOldest = 'Newest to Oldest',
  longestToShortest = 'Longest to Shortest',
  shortestToLongest = 'Shortest to Longest',
  firstPlayerToLastPlayer = 'First player to Last player',
  lastPlayerToFirstPlayer = 'Last player to First player'
}
