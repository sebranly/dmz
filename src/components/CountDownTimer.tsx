import * as React from 'react';
import { DEAD_DROP_HOURLY_RATE, REGULAR_HOURLY_RATE } from '../constants/game';
import { Timer, TimeUnit } from '../types';
import {
  calculateRemainingSeconds,
  convertSecondsToMoney,
  convertSecondsToTimeValue,
  convertTimerIndexToPlayerIndex,
  convertTimerIndexToPlayerTimerIndex,
  getEndTime
} from '../utils';
import { convertToTwoDigits, formatMoney, getPlayerColor, pluralize } from '../utils/display';

export interface CountDownTimerProps {
  className?: string;
  currentTimestamp: number;
  deleteTimer: () => void;
  timer: Timer;
}

const CountDownTimer: React.FC<CountDownTimerProps> = (props) => {
  const { className = '', currentTimestamp, deleteTimer, timer } = props;
  const { timerIndex } = timer;

  const remainingSeconds = calculateRemainingSeconds(timer, currentTimestamp);
  const endTime = getEndTime(timer);

  const {
    [TimeUnit.Hour]: hours,
    [TimeUnit.Minute]: minutes,
    [TimeUnit.Second]: seconds
  } = convertSecondsToTimeValue(remainingSeconds);

  const isFixedHours = hours === 0;
  const isFixedMinutes = isFixedHours && minutes === 0;
  const isFixedSeconds = isFixedMinutes && seconds === 0;

  const playerIndex = convertTimerIndexToPlayerIndex(timerIndex);
  const playerTimerIndex = convertTimerIndexToPlayerTimerIndex(timerIndex);
  const color = getPlayerColor(playerIndex);
  const playerTitle = `Player ${playerIndex + 1}`;
  const timerTitle = `Timer ${playerTimerIndex + 1}`;
  const showRibbon = currentTimestamp - timer.timestampStart <= 5;

  const classnamesPlayerColor = `color-${color}`;
  const classnamesPlayerTitle = `count-down-title ${classnamesPlayerColor}`;
  const classnamesComponent = `count-down ${className} border-color-${color} ribbon-container`;
  const classnamesMoney = `money-value ${classnamesPlayerColor}`;
  const classnamesTimerValue = 'timer-value';

  const items = [
    { value: hours, label: TimeUnit.Hour },
    { value: minutes, label: TimeUnit.Minute },
    { value: seconds, label: TimeUnit.Second }
  ].map((element: { value: number; label: string }) => {
    const { value, label: l } = element;
    const isFixed =
      (l === TimeUnit.Hour && isFixedHours) ||
      (l === TimeUnit.Minute && isFixedMinutes) ||
      (l === TimeUnit.Second && isFixedSeconds);

    const classnamesValue = isFixed ? `${classnamesPlayerColor} ${classnamesTimerValue}` : classnamesTimerValue;

    return (
      <li className="count-down-timer-element flex-child" key={l}>
        <span className={classnamesValue}>{convertToTwoDigits(value)}</span>
        <div className="count-down-timer-unit">{pluralize(l, value)}</div>
      </li>
    );
  });

  return (
    <div className={classnamesComponent}>
      {showRibbon && <div className="ribbon-child">NEW</div>}
      <div className={classnamesPlayerTitle}>{playerTitle}</div>
      <div className="count-down-title">{timerTitle}</div>
      <ul className="count-down-timer flex-container">{items}</ul>
      <div className="count-down-timer-money">
        <div className="money-title">End Time:</div> <div className={classnamesMoney}>{endTime}</div>
      </div>
      <div className="count-down-timer-money">
        <div className="money-title">Exfil Money:</div>{' '}
        <div className={classnamesMoney}>
          ${formatMoney(convertSecondsToMoney(remainingSeconds, REGULAR_HOURLY_RATE))}
        </div>
      </div>
      <div className="count-down-timer-money">
        <div className="money-title">Dead Drop:</div>{' '}
        <div className={classnamesMoney}>
          ${formatMoney(convertSecondsToMoney(remainingSeconds, DEAD_DROP_HOURLY_RATE))}
        </div>
      </div>
      <button className="remove-timer" onClick={() => deleteTimer()}>
        Delete this timer
      </button>
    </div>
  );
};

export { CountDownTimer };
