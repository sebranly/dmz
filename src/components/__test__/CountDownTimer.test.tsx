import * as React from 'react';
import { render } from '@testing-library/react';
import { CountDownTimer } from '../CountDownTimer';
import { Timer } from '../../types';

test('CountDownTimer', () => {
  const timer: Timer = { timerIndex: 0, durationSec: 62, timestampStart: 120 };
  const createComponent = () => <CountDownTimer currentTimestamp={123} deleteTimer={jest.fn()} timer={timer} />;

  const { container } = render(createComponent());
  expect(container.childNodes).toMatchSnapshot();
});
