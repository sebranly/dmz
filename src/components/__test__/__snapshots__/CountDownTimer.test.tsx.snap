// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountDownTimer 1`] = `
NodeList [
  <div
    class="count-down  border-color-green ribbon-container"
  >
    <div
      class="ribbon-child"
    >
      NEW
    </div>
    <div
      class="count-down-title color-green"
    >
      Player 1
    </div>
    <div
      class="count-down-title"
    >
      Timer 1
    </div>
    <ul
      class="count-down-timer flex-container"
    >
      <li
        class="count-down-timer-element flex-child"
      >
        <span
          class="color-green timer-value"
        >
          00
        </span>
        <div
          class="count-down-timer-unit"
        >
          hours
        </div>
      </li>
      <li
        class="count-down-timer-element flex-child"
      >
        <span
          class="color-green timer-value"
        >
          00
        </span>
        <div
          class="count-down-timer-unit"
        >
          minutes
        </div>
      </li>
      <li
        class="count-down-timer-element flex-child"
      >
        <span
          class="timer-value"
        >
          59
        </span>
        <div
          class="count-down-timer-unit"
        >
          seconds
        </div>
      </li>
    </ul>
    <div
      class="count-down-timer-money"
    >
      <div
        class="money-title"
      >
        End Time:
      </div>
       
      <div
        class="money-value color-green"
      >
        01:03 AM
      </div>
    </div>
    <div
      class="count-down-timer-money"
    >
      <div
        class="money-title"
      >
        Exfil Money:
      </div>
       
      <div
        class="money-value color-green"
      >
        $
        700
      </div>
    </div>
    <div
      class="count-down-timer-money"
    >
      <div
        class="money-title"
      >
        Dead Drop:
      </div>
       
      <div
        class="money-value color-green"
      >
        $
        500
      </div>
    </div>
    <button
      class="remove-timer"
    >
      Delete this timer
    </button>
  </div>,
]
`;
